modules/service_container_doctrine/src/Plugin/Discovery/AnnotatedClassDiscovery.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\service_container\Plugin\Discovery\AnnotatedClassDiscovery
+ */
+
+namespace Drupal\service_container_doctrine\Plugin\Discovery;
+
+use Doctrine\Common\Annotations\AnnotationRegistry;
+use Doctrine\Common\Annotations\SimpleAnnotationReader;
+use Doctrine\Common\Reflection\StaticReflectionParser;
+use Drupal\Component\Annotation\Reflection\MockFileFinder;
+use Drupal\Component\Plugin\Discovery\DiscoveryInterface;
+use Drupal\Component\Plugin\Exception\PluginNotFoundException;
+
+/**
+ * TODO
+ *
+ * This class cannot be tested as it relies on the existence of procedural code.
+ * @codeCoverageIgnore
+ */
+class AnnotatedClassDiscovery implements DiscoveryInterface {
+
+  /**
+   * The namespaces within which to find plugin classes.
+   *
+   * @var string[]
+   */
+  protected $pluginNamespaces;
+
+  /**
+   * The name of the annotation that contains the plugin definition.
+   *
+   * The class corresponding to this name must implement
+   * \Drupal\Component\Annotation\AnnotationInterface.
+   *
+   * @var string
+   */
+  protected $pluginDefinitionAnnotationName;
+
+  /**
+   * The doctrine annotation reader.
+   *
+   * @var \Doctrine\Common\Annotations\Reader
+   */
+  protected $annotationReader;
+
+  /**
+   * Constructs a new instance.
+   *
+   * @param string[] $plugin_manager_definition
+   *   (optional) An array of namespace that may contain plugin implementations.
+   *   Defaults to an empty array.
+   * @param string $plugin_definition_annotation_name
+   *   (optional) The name of the annotation that contains the plugin definition.
+   *   Defaults to 'Drupal\Component\Annotation\Plugin'.
+   */
+  function __construct($plugin_manager_definition, $plugin_definition_annotation_name = 'Drupal\Component\Annotation\Plugin') {
+    $directories = array();
+
+    foreach(module_list() as $module_name) {
+      $directories[] = DRUPAL_ROOT . '/' . drupal_get_path('module', $module_name) . '/' . trim($plugin_manager_definition['directory'], DIRECTORY_SEPARATOR);
+    }
+
+    $namespace = new \ArrayObject(array('Drupal\\' . $plugin_manager_definition['owner'] . '\\Plugin' => $directories));
```

Comment:
That is wrong and only works by accident.

This needs to be a data structure like:

```
Drupal\[module_name]\[directory=>class_namespace]\ => [
  DRUPAL_ROOT . '/' . drupal_get_path('module', $module_name) . '/' . 'src/' . $plugin_manager_definition['directory']
]
```

To convert from directory to namespace, use:

$directory = str_replace(DIRECTORY_SEPARATOR, '\\', $directory) . '\\' : '';